git-discover/src/upwards.rs

```diff
@@ -75,7 +83,29 @@ pub(crate) mod function {
         };
 
         let mut cursor = dir.clone().into_owned();
+
+        let max_height = if !ceiling_dirs.is_empty() {
+            // Ceiling directory discovery requires us to canonicalize the path
+            // TODO: this works nonetheless. Is a test missing? Or does 'absolutize()' do the job already?
+            // is_canonicalized = true;
+            // cursor = cursor
+            //     .canonicalize()
+            //     .map_err(|_| Error::InaccessibleDirectory { path: cursor })?;
```

Comment:
Without this, this was breaking the `upwards::from_dir_with_dot_dot` test (the `is_empty` causes it not to fail now).

Reply:
Thinking about it, the prefix check would only be sensible on a canonicalized directory _or_ if all paths are relative to the same root. The latter is something that could also be documented as requirement. Maybe there could be an assertion for this as well.

The reason I am not keen to canonicalize by default is that it alters paths that the user as provided, which might elongate them which makes it uncomfortable to read.

It's a bit hard for me to reason about the roots of paths in this portion of the code unfortunately, but it should always be possible to pin requirements with tests.

Reply:
> Thinking about it, the prefix check would only be sensible on a canonicalized directory or if all paths are relative to the same root. The latter is something that could also be documented as requirement. Maybe there could be an assertion for this as well.

The `GIT_CEILING_DIRECTORIES` env variable for this requires absolute paths, and I think non-absolute paths would be too brittle to be useful in practice. Or do you mean same root as in something like drives on Windows?

> The reason I am not keen to canonicalize by default is that it alters paths that the user as provided, which might elongate them which makes it uncomfortable to read.

Would not setting the cursor be acceptable? Though that would mean calling `canonicalize` twice in some circumstances.

Reply:
> The GIT_CEILING_DIRECTORIES env variable for this requires absolute paths, and I think non-absolute paths would be too brittle to be useful in practice. Or do you mean same root as in something like drives on Windows?

Now that I was about to write about this I realize that what's merged by me still has shortcomings, so I am going to fix this on main.
The current implementation kicks the can down the street and tells the caller to assure paths are normalized according to their requirements.